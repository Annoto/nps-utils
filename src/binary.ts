import * as path from "path";

/**
 * Get the path to one of the bin scripts exported by a package
 * @param packageName - name of the npm package
 * @param binName - name of the script. Defaults to packageName
 * @returns path, relative to process.cwd()
 */
export function getBin(packageName: string, binName = packageName): string {
    const packagePath = require.resolve(`${packageName}/package.json`);
    const concurrentlyDir = path.dirname(packagePath);
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    let { bin: binRelativeToPackage } = require(packagePath);

    if (typeof binRelativeToPackage === "object") {
        binRelativeToPackage = binRelativeToPackage[binName];
    }

    const fullBinPath = path.join(concurrentlyDir, binRelativeToPackage);

    return path.relative(process.cwd(), fullBinPath);
}

/**
 * Get the path to one of the bin scripts exported by a package
 * @param packageName - name of the npm package
 * @param binName - name of the script. Defaults to packageName
 * @returns path, relative to process.cwd()
 */
export function runBin(packageName: string, binName = packageName): string {
    return `node ${getBin(packageName, binName)}`;
}
