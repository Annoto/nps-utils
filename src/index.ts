export * from "./series";

export * from "./binary";

export * from "./rimraf";

export * from "./copy";

export * from "./crossEnv";

export * from "./isDocker";

export * from "./concurrent";
