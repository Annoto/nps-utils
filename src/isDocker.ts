import isRunningInsideDocker from "is-docker";

/**
 * Check if the process is running inside a Docker container
 * @return - is running inside docker.
 */
export function isDocker(): boolean {
    return isRunningInsideDocker();
}
