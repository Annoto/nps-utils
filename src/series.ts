import { quoteScript } from "./utils/quoteScript";

/**
 * Accepts any number of scripts, filters out any
 * falsy ones and joins them with ' && '
 * @param scripts - Any number of strings representing commands
 * @example
 * // returns 'eslint && jest && webpack --env.production'
 * series('eslint', 'jest', 'webpack --env.production')
 * @return - the command that will execute the given scripts in series
 */
export function series(...scripts: string[]): string {
    return scripts.filter(Boolean).join(" && ");
}

/**
 * Accepts any number of nps script names, filters out
 * any falsy ones, prepends `nps` to them, and passes
 * the that to `series`
 * @param scriptNames - the script names to run
 * // returns 'nps lint && nps "test --coverage" && nps build'
 * series.nps('lint', 'test --coverage', 'build')
 * @return - the command that will execute the nps scripts in series
 */
export function seriesNPS(...scriptNames: string[]): string {
    return series(
        ...scriptNames
            .filter(Boolean)
            .map((scriptName) => scriptName.trim())
            .filter(Boolean)
            .map((scriptName) => `nps ${quoteScript(scriptName)}`),
    );
}
