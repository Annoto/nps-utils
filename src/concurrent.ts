import { runBin } from "./binary";
import { quoteScript } from "./utils/quoteScript";

const defaultColors = [
    "bgBlue.bold",
    "bgMagenta.bold",
    "bgGreen.bold",
    "bgBlack.bold",
    "bgCyan.bold",
    "bgRed.bold",
    "bgWhite.bold",
    "bgYellow.bold",
    // TODO: add more colors that look good?
];

/**
 * Escape a string so the shell expands it to the original.
 * @param - as accepted by any-shell-escape; arrays will
 * yield multiple arguments in the shell
 * @returns ready to pass to shell
 */
function shellEscape(arg: string | string[]): string {
    // lazily require for perf :)
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return require("any-shell-escape")(arg);
}

interface ConcurrentScript {
    script: string;
    color?: string;
}

/**
 * Generates a command that uses `concurrently` to run
 * scripts concurrently. Adds a few flags to make it
 * behave as you probably want (like --kill-others-on-fail).
 * In addition, it adds color and labels where the color
 * can be specified or is defaulted and the label is based
 * on the key for the script.
 * @param scripts - the scripts to run
 *   note: this function filters out falsy values :)
 * @example
 * // returns a bit of a long script that can vary slightly
 * // based on your environment... :)
 * concurrent({
 *   lint: {
 *     script: 'eslint .',
 *     color: 'bgGreen.white.dim',
 *   },
 *   test: 'jest',
 *   build: {
 *     script: 'webpack'
 *   }
 * })
 * @return - the command to run
 */
export function concurrent(scripts: Record<string, ConcurrentScript | string>): string {
    interface ConcurrentlyScripts {
        colors: string[];
        scripts: string[];
        names: string[];
    }

    const concurrentlyScripts: ConcurrentlyScripts = Object.keys(scripts).reduce(
        (previousValue, scriptName, currentIndex) => {
            let scriptObj = scripts[scriptName];

            if (!scriptObj) return previousValue;

            if (typeof scriptObj === "string") scriptObj = { script: scriptObj };

            const { script, color = defaultColors[currentIndex % defaultColors.length] } = scriptObj;

            if (!script) return previousValue;

            previousValue.names.push(scriptName);
            previousValue.colors.push(color);
            previousValue.scripts.push(script);

            return previousValue;
        },
        {
            colors: [],
            names: [],
            scripts: [],
        } as ConcurrentlyScripts,
    );

    const { colors, names, scripts: processedScripts } = concurrentlyScripts;

    const flags = [
        "--kill-others-on-fail",
        `--prefix-colors "${colors.join(",")}"`,
        '--prefix "[{name}]"',
        `--names "${names.join(",")}"`,
        shellEscape(processedScripts),
    ];

    const concurrently = runBin("concurrently");

    return `${concurrently} ${flags.join(" ")}`;
}

/**
 * Accepts any number of nps script names, filters out
 * any falsy ones, prepends `nps` to them, and passes
 * the that to `concurrent`
 * @param scriptNames - the script names to run
 * @example
 * // will basically return `nps lint & nps "test --coverage" & nps build`
 * // but with the concurrently command and relevant flags to make
 * // it super awesome with colors and whatnot. :)
 * concurrent.nps('lint', 'test --coverage', 'build')
 * @return the command to run
 */
export function concurrentNPS(...npsScripts: string[]): string {
    function reduceNPSScripts(
        previousVal: Record<string, ConcurrentScript>,
        currentVal: ConcurrentScript | undefined,
    ): Record<string, ConcurrentScript> {
        if (!currentVal) return previousVal;
        const { color, script } = currentVal;
        const [name] = script.split(" ");

        previousVal[name] = { script: `nps ${quoteScript(script.trim())}`, color };

        return previousVal;
    }

    function mapNPSScripts(script: string, index: number): ConcurrentScript | undefined {
        const color = defaultColors[index];

        if (!Boolean(script)) return undefined;

        if (typeof script === "string") return { script, color };

        return Object.assign({ color }, script);
    }

    return concurrent(npsScripts.map(mapNPSScripts).reduce(reduceNPSScripts, {} as Record<string, ConcurrentScript>));
}
