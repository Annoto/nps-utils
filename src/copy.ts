import { runBin } from "./binary";

/**
 * Gets a script that uses the cpy-cli binary. cpy-cli
 * is a dependency of nps-utils, so you don't need to
 * install it yourself.
 * @param args - args to pass to cpy-cli
 *   learn more from http://npm.im/cpy-cli
 * @return - the command with the cpy-cli binary
 */
export function copy(args: string): string {
    return `${runBin("cpy-cli", "cpy")} ${args}`;
}
