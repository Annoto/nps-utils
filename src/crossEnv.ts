import { runBin } from "./binary";

/**
 * Gets a script that uses the cross-env binary. cross-env
 * is a dependency of nps-utils, so you don't need to
 * install it yourself.
 * @param args - args to pass to cross-env
 *   learn more from http://npm.im/cross-env
 * @return - the command with the cross-env binary
 */
export function crossEnv(args: string): string {
    return `${runBin("cross-env")} ${args}`;
}
