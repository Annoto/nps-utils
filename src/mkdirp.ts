import { runBin } from "./binary";

/**
 * Gets a script that uses the mkdirp binary. mkdirp
 * is a dependency of nps-utils, so you don't need to
 * install it yourself.
 * @param args - args to pass to mkdirp
 *   learn more from http://npm.im/mkdirp
 * @return - the command with the mkdirp binary
 */
export function mkdirp(args: string): string {
    return `${runBin("mkdirp")} ${args}`;
}
