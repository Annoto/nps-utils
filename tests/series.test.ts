import { series, seriesNPS } from "../src/series";

describe("Test Series Helpers...", () => {
    it("Should combine multiple commands into one command", () => {
        const args = ["echo hey", null, "echo hi", undefined, "echo there"] as string[];

        expect(series(...args)).toMatchSnapshot();
    });

    it("Should combine multiple nps commands into one command", () => {
        const args = [
            "test",
            false,
            "lint.src",
            null,
            "lint.scripts --cache",
            undefined,
            " ",
            "build --fast",
        ] as string[];

        expect(seriesNPS(...args)).toMatchSnapshot();
    });
});
