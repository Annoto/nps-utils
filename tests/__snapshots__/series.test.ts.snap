// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Series Helpers... Should combine multiple commands into one command 1`] = `"echo hey && echo hi && echo there"`;

exports[`Test Series Helpers... Should combine multiple nps commands into one command 1`] = `"nps test && nps lint.src && nps \\"lint.scripts --cache\\" && nps \\"build --fast\\""`;
